src/containers/TrackList.js

```diff
@@ -1,9 +1,42 @@
 import { connect } from 'react-redux';
 import { getNewTracks, startSession } from '../actions';
-import TrackListView from '../components/TrackListView';
+import ViewableTrackList from '../components/ViewableTrackList';
+import _ from 'lodash';
+
+const filterByTrackName = (trackList, filters) => {
+  return filters.byTrackName === ''
+    ? trackList
+    : _.filter(trackList, o => {
+        return o.title
+          .toLowerCase()
+          .includes(filters.byTrackName.trimLeft().toLowerCase());
+      });
+};
+
+const filterByArtistName = (trackList, filters) => {
+  return filters.byArtistName === ''
+    ? trackList
+    : _.filter(trackList, o => {
+        return o.user.username
+          .toLowerCase()
+          .includes(filters.byArtistName.trimLeft().toLowerCase());
+      });
+};
+
+const FILTER_LIST = [filterByTrackName, filterByArtistName];
+
+const filterTrackList = (trackList, filters) => {
+  return _.intersection(..._.map(FILTER_LIST, f => f(trackList, filters)));
```

Comment:
This is the right concept, but a less efficient data structure. Logically you want the intersection of sets, but algorithmically it's better to apply a series of filters (each on the result of the previous).

Filters shrink the result at each stage and can be chained. Sets require each whole set to be present in memory before calculating the intersection.